src/operator/controllers/poduserpassword/db_credentials_pod_reconciler.go

```diff
@@ -79,60 +79,120 @@ func (e *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
+func (e *Reconciler) ensurePodUserAndPasswordSecret(ctx context.Context, pod *v1.Pod, secretName string) (error, string) {
 	log := logrus.WithFields(logrus.Fields{"pod": pod.Name, "namespace": pod.Namespace})
-	err := e.client.Get(ctx, types.NamespacedName{Namespace: pod.Namespace, Name: secretName}, &v1.Secret{})
+	secret := v1.Secret{}
+	err := e.client.Get(ctx, types.NamespacedName{Namespace: pod.Namespace, Name: secretName}, &secret)
 	if apierrors.IsNotFound(err) {
 		log.Debug("Creating user-password credentials secret for pod")
-		password, err := createServicePassword()
-		if err != nil {
-			return errors.Wrap(err)
-		}
-		clusterID, err := clusterutils.GetClusterUID(ctx)
+		password, err := databaseconfigurator.GenerateRandomPassword()
 		if err != nil {
-			return errors.Wrap(err)
+			return errors.Wrap(err), ""
 		}
 
-		username := clusterutils.BuildHashedUsername(serviceName, pod.Namespace, clusterID)
-		pgUsername := clusterutils.KubernetesToPostgresName(username)
+		databaseUsername := pod.Annotations[databaseconfigurator.DatabaseUsernameAnnotation]
 
-		secret := buildUserAndPasswordCredentialsSecret(secretName, pod.Namespace, pgUsername, password)
+		secret := buildUserAndPasswordCredentialsSecret(secretName, pod.Namespace, databaseUsername, password)
 		log.WithField("secret", secretName).Debug("Creating new secret with user-password credentials")
 		if err := e.client.Create(ctx, secret); err != nil {
-			return errors.Wrap(err)
+			return errors.Wrap(err), ""
 		}
-		return nil
+		return nil, password
 	}
 
 	if err != nil {
-		return errors.Wrap(err)
+		return errors.Wrap(err), ""
 	}
 	log.Debug("Secret exists, nothing to do")
+	return nil, string(secret.Data["password"])
+}
+
+func (e *Reconciler) ensurePasswordInDatabases(ctx context.Context, pod v1.Pod, password string) error {
+	databases := strings.Split(pod.Annotations[databaseconfigurator.DatabaseAccessAnnotation], ",")
+	username := pod.Annotations[databaseconfigurator.DatabaseUsernameAnnotation]
+	pgServerConfigs := otterizev1alpha3.PostgreSQLServerConfigList{}
+	err := e.client.List(ctx, &pgServerConfigs)
+	if err != nil {
+		return errors.Wrap(err)
+	}
+
+	mysqlServerConfigs := otterizev1alpha3.MySQLServerConfigList{}
+	err = e.client.List(ctx, &mysqlServerConfigs)
+	if err != nil {
+		return errors.Wrap(err)
+	}
+
+	for _, database := range databases {
+		dbconfigurator, found, err := e.createDBConfigurator(ctx, database, mysqlServerConfigs.Items, pgServerConfigs.Items)
+		if err != nil {
+			return errors.Wrap(err)
+		}
+		if !found {
+			logrus.Warningf("Missing database server config for db: %s", database)
+			e.recorder.Eventf(&pod, v1.EventTypeWarning, ReasonEnsuringDatabasePasswordFailed,
+				"Failed to ensure database password in %s. Missing database server config", database)
+			continue
+		}
+		defer dbconfigurator.CloseConnection(ctx)
```

Comment:
`defer` in for loops has some caveats in it: 
![image](https://github.com/otterize/credentials-operator/assets/10258630/43c609c0-50a0-4274-9c8c-dad1958d7e58)

what's certain is, it'll only be called before the entire function exits, and not at the end of the for loop iteration, which is wasteful. 
The best way to handle it is put the entire loop iteration in a dedicated function (including the defer): 

```
for _, database := range databases {
    found, err := e.ensurePasswordInDatabaseInstance(ctx, database, mysqlServerConfigs.Items, pgServerConfigs.Items, username, password)
}


func (r *Reconciler) ensurePasswordInDatabaseInstance(ctx, ...) {
    dbConfigurator, found, err := r.createDBConfigurator(...)
    if err != nil {...}
    if !found {return false, nil}

    defer dbConfigurator.CloseConnection() // this way it's guaranteed to be called as soon as this internal function exists, which is as soon as it's not needed
    return dbConfigurator.AlterUserPassword(...)   
}
```